unifying 2D and 3D messages

procopiostein opened the pull request:
IMO I think we gain in simplicity by reducing the number of messages by merging, when possible, 2D and 3D msgs. 
A pose can be interpreted as 2D or 3D according to the value on the z field. 
For example, in many cases, wheeled robots in ROS publish their pose as a geometry_msgs/Pose, (inherently 3D) although they are always at the ground level.

tfoote commented on msg/BoundingBox.msg:
```diff
@@ -1,7 +1,7 @@
 # size of the bounding box
-geometry_msgs/Vector3 size
\ No newline at end of file
+geometry_msgs/Point size
```
Why change this to a Point?

This should be documented how to resolve this. What is it relative to? The pose of the `pose`?

procopiostein replied:
I reverted to the original Vector type, and added a better description.

tfoote commented on msg/BoundingBox.msg:
```diff
@@ -1,7 +1,7 @@
 
-# position and rotation of the bounding box
+# position and orientation of the bounding box center
 geometry_msgs/Pose pose
```
Not in this PR but related to the other comment. This should probably be more considered an 'origin' or something instead of echoing the datatype.

procopiostein replied:
I improved the description.

Kukanani commented:
+1 on combining ObjectHypothesis. The `DetectionXD` messages will also  have to be updated to point to the new message type.

Thanks for the PR!

procopiostein commented:
Agreed @tfoote , I have not thought about that. I reverted the related commit and tried to improved the description and fields. Also changed the datatype for the size of a 2D BB to uint32, as it should be in pixels.

@Kukanani  I updated the DetectionXD to use the generic ObjectHypothesisWithPose

mintar commented on msg/BoundingBox2D.msg:
```diff
@@ -1,9 +1,9 @@
-# The x/y position and (optional) rotation of the bounding box center.
-geometry_msgs/Pose2D pose
+# The 2D position (in pixels) and orientation of the bounding box origin.
+geometry_msgs/Pose2D origin
```
Honest question: Is it common to have rotated bounding boxes in 2D classification? I've only seen axis-aligned bounding boxes so far. If so, we should replace the Pose2D with a "Point2D" (which doesn't exist yet), preferably with uint32 coordinates instead of float. If (almost) nobody uses the rotation, many people won't implement it correctly.

Kukanani replied:
Rotated 2D bounding boxes often is used for grasp generators and human annotations. Since the bounding boxes will probably move to geometry_msgs, I think that a generic rotated version should be best.

mintar commented on msg/BoundingBox2D.msg:
```diff
@@ -1,9 +1,9 @@
+# The size (in pixels) of the bounding box surrounding the object relative 
+# to the pose of its origin
+uint32 size_x
+uint32 size_y
```
Related to my previous comment: If we allow the bounding box origin to be given as floats with rotation, we should also allow floats for the size. The unit should still be pixels. However, I would prefer to remove the rotation and use uint32 everywhere (bounding box origin + size).

procopiostein replied:
I agree with both your comments.

Kukanani replied:
Sub-pixel bounding boxes are possible when using machine learning-based object detectors, but it's worth asking if that information is worth storing.

I'll also note that without floats for `origin`, it becomes difficult to accurately position an even-sized bounding box (if the BB is 2x2, where is its center?)

mintar replied:
I think that by `origin`, @procopiostein meant something like the top-left corner, not the center. This is why a 2x2 BB is not a problem: it's the box between (origin.x, origin.y) and (origin.x + size_x, origin.y + size_y).

mintar commented:
Let's continue the discussion in #5.
